### DISCUSSION TITLE
Pause and resume notification option

### DISCUSSION THREAD

**Discussion by crwagner89** [berrystephenw/OctoText#9] [Ideas]
I've been using the plug in and it seems to be working well so far, but I had some feedback. My intention is to put filament runout sensors on all my printers and set them up to pause their prints when the sensors detects the absence of filament. Once the printer is paused I want to receive a text notification so I know to swap in new filament.  I went ahead and modified the OctoText code to send notifications when it pauses and resumes, but I imagine other people would use it this way so it might be useful as a standard option. 

The other thing I noticed is the formatting seems off in the messages that are sent in normal operation (not the test message). They all start with "None /" and then it's followed with the title and description. What is supposed to be in place of the "None /" at the beginning of the message? Would it be possible to replace this with the login name of the octoprint profile? I have multiple printers so it would be nice to know which one the message is coming from. An example would be:

Prusa
The print has paused, check filament
Spatula.gcode has been paused locally

Just some feedback, overall it's working great. Thanks for taking the time to make this.

**Comment by berrystephenw** [comment]
Thank you for the feedback, I appreciate it. 

As for the formatting of the messages I think there will be issues depending on the type of phone (I'm assuming you are talking about receiving texts). I can certainly change the formatting but I don't have a good way of testing it. Let me know what phone you are using and what app you are using - maybe there is something I can try. I've thought about incorporating the printer name in the message not so much the user name... I'll put it on the next revision to look at!

As for your filament sensors, I would need to know how it is integrated into the system now. Is it a gcode command sent from the printer? I have a Prusa as well and I could test at least that printer but I can see how that might be a support nightmare if every printer does it differently. Let me look into that.

Thanks!

**Reply by crwagner89** [reply]
Just to be clear, I only meant it would be nice to have pause and resume as standard event options that would trigger a message being sent. The filament sensor is just a switch plugged into an input pin on the Pi. I plan on handling the filament sensor with a separate plugin, like "Filament sensor reloaded plugin". This plug in would simply pause the print when the filament sensor triggers. OctoText would then send a message when the printer paused, not really caring why it paused. That seems like a straight-forward way to accomplish what I want (at least in my head). 

Having the printer names in the text message would definitely be a useful thing. For reference
 I'm sending it to my phone it via outlook, phone is a pixel 4a on T-mobile. 

![Prusa OctoText example](https://user-images.githubusercontent.com/81662750/113174642-944ee180-9218-11eb-93db-3c66a1780473.png)

**Reply by berrystephenw** [reply]
Got it. Right now the events that I receive from Octoprint do not include pause and resume. I'm pretty sure that I can figure out how that could be handled, and I see how that would be useful. Thanks! I'll look into the text message formatting - the screen shot is helpful since the code takes a different path depending if the webcam is being used or not. If you have a cam, could you enable that and see if the "None /" messages are different?

**Reply by crwagner89** [reply]
Oddly when I have the webcam enabled I get an error saying "Exception while talking to your mail server, check your SMTP settings". Here's a sample of the log:

2021-03-31 17:43:30,510 - octoprint.plugins.OctoText - INFO - Enable webcam setting True
2021-03-31 17:43:30,511 - octoprint.plugins.OctoText - INFO - filename for image: test-cvRarQgbZcXiAVvG.jpg
2021-03-31 17:43:30,511 - octoprint.plugins.OctoText - INFO - Webcam URL is: /webcam/?action=stream, Snapshot URL is: http://127.0.0.1:8080/?action=snapshot
2021-03-31 17:43:30,554 - octoprint.plugins.OctoText - INFO - Webcam tempfile /tmp/tmp2h_n38y8.jpg
2021-03-31 17:43:30,556 - octoprint.plugin - ERROR - Error while calling plugin OctoText
Traceback (most recent call last):
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/plugin/__init__.py", line 271, in call_plugin
    result = getattr(plugin, method)(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint/util/__init__.py", line 1890, in wrapper
    return f(*args, **kwargs)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_OctoText/__init__.py", line 455, in on_event
    status = self._send_message_with_webcam_image(title, description)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_OctoText/__init__.py", line 245, in _send_message_with_webcam_image
    self._process_snapshot(tempFile)
  File "/home/pi/oprint/lib/python3.7/site-packages/octoprint_OctoText/__init__.py", line 320, in _process_snapshot
    self._logger.info("Running: {}".format(" ".join(ffmpeg_command)))
TypeError: sequence item 3: expected str instance, _TemporaryFileWrapper found

**Reply by crwagner89** [reply]
Looks like "None" actually comes from the subject line of the email. 

![Capture](https://user-images.githubusercontent.com/81662750/113184117-ccf3b880-9222-11eb-907f-7073a0eac975.PNG)

**Reply by berrystephenw** [reply]
Yes, someone else pointed out a webcam problem that was similar. Do you rotate your webcam or change any of these settings:
![image](https://user-images.githubusercontent.com/44931130/113207610-cd9a4800-923e-11eb-80c7-a347097db788.png)

**Comment by mpd352** [comment]
Just want to add on to these comments.

Octotext is something I've been looking for.  Love it.  Thank you very much!

Like crwagner89, I have multiple octoprint instances running, so having the printer name in the text message would be great.  I also wanted a text app primarily to notify me if the printer paused due to a filament sensor being tripped, so I also vote for printer paused events as a trigger.

Finally, this is a low priority, but all of my cameras are flipped, so getting the bug that stops Octotext from sending web camera snapshots on flipped cameras would be nice.

I'll generate a log of what messages are sent between Octoprint and the printer when the filament sensor is tripped, and send it to you if that would help.

**Comment by mpd352** [comment]
I made a couple of dummy runs on my Ender 3.  Octotext does detect that the print has failed and sends a message to that effect:

FRM:mpd352@cox.net
SUBJ:None
MSG:
Print Fail 3
Print Fail 3 cancelled

It isn’t, however, specific to the filament sensor detecting no filament.

Messages coming from Octoprint do specify that the filament sensor has detected no filament. 

I’ve no idea how you do to the magic to create these apps, but I’m betting when you have a chance you’ll be able to figure out how to capture that message and give us guys using filament sensors specific messages that the filament has run out.

Kind of a holy Grail for some of us because we have rolls and rollsof very short pieces of filament that we would like to use, and it would be nice for the Octoprint to nudge us when it’s time to swap filament.

The log is attached, if you search it for “Octotext” and/or “filament manager”  you’ll see that the messages are intertwined with each other where the printer is saying that the filament sensor detects no filament, and in the next line you can see octotext sending the print failed message.

Hope this helps when you get to this point in the requested features queue.

[octoprint.log](https://github.com/berrystephenw/OctoText/files/6251409/octoprint.log)

**Comment by berrystephenw** [comment]
I'm glad octotext is useful for you! 
Do you have a plugin called 'filamentreload' ? It might be builtin but I haven't seen that before.

**Comment by mpd352** [comment]
I use this one:

***@***.***D727E3.D479CAD0]

It’s no longer being supported, but still works.  There are others, and if you find one of those are a better fit, I’ll switch.  I figure I’m going to have to find a new plug in eventually since it’s originator dropped off the face of the earth.

Thank you,

Kevin Conde
Peoria, AZ

**Comment by mpd352** [comment]
This is the plug-in page, should have sent it with my first reply.

https://plugins.octoprint.org/plugins/filament_sensor_reloaded/

Thank you,

Kevin Conde
Peoria, AZ

**Reply by berrystephenw** [reply]
Great thank you - I'll look through this and other ways of detecting a filament change. My Prusa has a built in sensor and now I'm wondering how that is reported...

**Comment by berrystephenw** [comment]
Ok so the rc branch has print paused and resumed modifications added, as well as fixing the "none" subject (a weird bug) and adding the printer name into the message body. If anyone is interested in trying these new features please see this post #34 for how to do that. Once there has been some test time on this version I will merge it with main and release it to everyone. Thanks for reporting these problems and giving me feedback!
